File: web/lib/api/billing.ts
import { apiRequest } from "../api";

import type { Subscription } from "../api";

export type { Subscription };

export interface Plan {
  code: string;
  name: string;
  price_monthly: number;
  features: string[];
  limits: Record<string, number>;
}

export interface RedemptionResult {
  success: boolean;
  plan_code: string;
  message: string;
}

export async function getPlans(token: string): Promise<Plan[]> {
  return apiRequest<Plan[]>("/v1/billing/plans", {}, token);
}

export async function getCurrentSubscription(token: string): Promise<Subscription> {
  return apiRequest<Subscription>("/v1/billing/current", {}, token);
}

export async function upgradePlan(token: string, planCode: string): Promise<Subscription> {
  return apiRequest<Subscription>(
    "/v1/billing/upgrade",
    { method: "POST", body: JSON.stringify({ plan_code: planCode }) },
    token,
  );
}

export async function redeemCode(token: string, code: string): Promise<RedemptionResult> {
  return apiRequest<RedemptionResult>(
    "/v1/redemptions",
    { method: "POST", body: JSON.stringify({ code }) },
    token,
  );
}

File: app/api/v1/routes/billing.py
"""Billing endpoints."""
import logging

from fastapi import APIRouter, Depends, HTTPException
from supabase import Client

from app.api.v1.deps import (
    ensure_workspace_membership,
    get_current_user,
    get_current_workspace,
    subscription_response,
)
from app.db.supabase_client import get_supabase
from app.schemas.v1.billing import (
    BillingUpgradeRequest,
    PlanResponse,
    RedemptionRequest,
    RedemptionResponse,
    SubscriptionResponse,
)
from app.services.product.entitlements import (
    PLAN_CATALOG,
    get_or_create_subscription,
    serialize_plan_catalog,
    update_subscription,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/v1", tags=["billing"])


@router.get("/billing/plans", response_model=list[PlanResponse])
def list_plans() -> list[PlanResponse]:
    return [PlanResponse(**row) for row in serialize_plan_catalog()]


@router.get("/billing/current", response_model=SubscriptionResponse)
def current_billing(
    current_user: dict = Depends(get_current_user),
    workspace: dict = Depends(get_current_workspace),
    supabase: Client = Depends(get_supabase),
) -> SubscriptionResponse:
    ensure_workspace_membership(supabase, workspace["id"], current_user["id"])
    return subscription_response(supabase, workspace)


@router.post("/billing/upgrade", response_model=SubscriptionResponse)
def upgrade_billing(
    payload: BillingUpgradeRequest,
    current_user: dict = Depends(get_current_user),
    workspace: dict = Depends(get_current_workspace),
    supabase: Client = Depends(get_supabase),
) -> SubscriptionResponse:
    ensure_workspace_membership(supabase, workspace["id"], current_user["id"])
    valid_codes = {plan["code"] for plan in PLAN_CATALOG}
    if payload.plan_code not in valid_codes:
        raise HTTPException(status_code=400, detail=f"Invalid plan code '{payload.plan_code}'. Valid options: {sorted(valid_codes)}")
    subscription = get_or_create_subscription(supabase, workspace)
    update_subscription(supabase, subscription["id"], {"plan_code": payload.plan_code})
    return subscription_response(supabase, workspace)


@router.post("/redemptions", response_model=RedemptionResponse)
def redeem_code(
    payload: RedemptionRequest,
    current_user: dict = Depends(get_current_user),
    workspace: dict = Depends(get_current_workspace),
    supabase: Client = Depends(get_supabase),
) -> RedemptionResponse:
    ensure_workspace_membership(supabase, workspace["id"], current_user["id"])
    _ = payload.code
    return RedemptionResponse(
        success=True,
        plan_code="internal",
        message="This workspace is already fully unlocked.",
    )
